### ISSUE TITLE
`Failed to find scope for next step`

### ISSUE THREAD

**Issue by epellizzer** [typelevel/fs2#1624]
Hi.

I have written a `Pipe2` that computes a diff between two sorted streams, but I run into an error at the end the execution.

```plain
Fail to find scope for next step: current: Token(61eefabb), step: Step(FreeC.Bind(FreeC.Bind(FreeC.Pure(()))),Some(Token(8798dd1)))
java.lang.Throwable: Fail to find scope for next step: current: Token(61eefabb), step: Step(FreeC.Bind(FreeC.Bind(FreeC.Pure(()))),Some(Token(8798dd1)))
	at fs2.internal.Algebra$.$anonfun$compileLoop$5(Algebra.scala:294)
	at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(IORunLoop.scala:139)
	at cats.effect.internals.IORunLoop$RestartCallback.signal(IORunLoop.scala:355)
	at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:376)
	at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:316)
	at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(IORunLoop.scala:136)
	at cats.effect.internals.IORunLoop$RestartCallback.signal(IORunLoop.scala:355)
	at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:376)
	at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:316)
	at cats.effect.internals.IOShift$Tick.run(IOShift.scala:36)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1425)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
```

To ease everyone's life, I have created [this GitHub project](https://github.com/epellizzer/showcase-fs2-diff) that contains the commented code as well as the unit tests that show the problem.

It was the first time I wrote a `Pipe2`, so I hope my implementation is not too naive, but even so, FS2 should probably not react that way.

Note that my implementation seems to show another problem with FS2 (a memory leak), but for this I will create another issue.

**Comment by epellizzer** [comment]
I have ran the JVM with JFR recording enabled and pushed the resulting recording in the GitHub project I mentioned above.

**Comment by epellizzer** [comment]
I tried to write my implementation differently and discovered `StepLeg`. It seems I'm exactly in the case where I should use it.

I get wrong results, so I obviously have a bug, but there is no more exception.

If step legs are mandatory when pulling two streams at a time, what could be done to prevent naive implementations like the one I came up with ? It seems nearly impossible to protect against it at compile time.

I leave it to you guys to decide what you want to do with this issue. I think I can reasonably consider things are clear for me.
